File: leetcode/323.number-of-connected-components-in-an-undirected-graph.py
class Solution(object):
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        graph = {i:[] for i in range(n)}
        
        for pair in edges:
            graph[pair[0]].append(pair[1])
            graph[pair[1]].append(pair[0])
            
        visited = set()
        count = 0
        for i in range(n):
            if i not in visited:
                count += 1
                self.dfs(graph, i, visited)
                
        return count
                
    def dfs(self, graph, start, visited):
        queue = [start]
        visited.add(start)
        
        while len(queue) > 0:
            tmp = queue.pop(0)
            
            for neighbor in graph[tmp]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

File: leetcode/323.number-of-connected-components-in-an-undirected-graph.cpp
class Solution {
private:
    void bfs(unordered_map<int, vector<int>>& graph, int start, unordered_set<int>& visited)
    {        
        queue<int> q;
        q.push(start);
        visited.insert(start);
        
        while(!q.empty())
        {
            int tmp = q.front();
            q.pop();
            
            for (auto neighbor: graph[tmp])
            {
                if (visited.count(neighbor) == 0)
                {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }
    }
public:
    int countComponents(int n, vector<pair<int, int>>& edges) {
        
        unordered_map<int, vector<int>>graph;
        unordered_set<int>visited;
        int count = 0;
        
        for (auto i = 0; i < edges.size(); ++i)
        {
            graph[edges[i].first].push_back(edges[i].second);
            graph[edges[i].second].push_back(edges[i].first);
        }
        
        for (auto i = 0; i < n; ++i)
        {
            if (visited.count(i) == 0)
            {
                count++;
                bfs(graph, i, visited);
            }
        }
        
        return count;
    }
};